e2e/appActions.js

```diff
@@ -405,19 +406,82 @@ async function selectInspectorTab(page, name) {
     }
 }
 
+/**
+* Waits and asserts that all plot series data on the page
+* is loaded and drawn.
+* @param {import('@playwright/test').Page} page
+*/
+async function waitForPlotsToRender(page) {
+    const plotLocator = page.locator('.gl-plot');
+    for (const plot of await plotLocator.all()) {
+        await expect(plot).toHaveClass(/js-series-data-loaded/);
+    }
+}
+
+/**
+ * @typedef {Object} PlotPixel
+ * @property {number} startIndex The index of the first value in the pixel
+ * @property {number} endIndex The index of the last value in the pixel
+ * @property {string} value The rgba string value of the pixel
+ */
+
+/**
+ * Wait for all plots to render and then retrieve and return an array
+ * of canvas plot pixel data (RGBA values).
+ * @param {import('@playwright/test').Page} page
+ * @param {string} canvasSelector The selector for the canvas element
+ * @return {Promise<PlotPixel[]>}
+ */
+async function getCanvasPixels(page, canvasSelector) {
+    const getTelemValuePromise = new Promise(resolve => page.exposeFunction('getCanvasValue', resolve));
+    const canvasHandle = await page.evaluateHandle((canvas) => document.querySelector(canvas), canvasSelector);
+    const canvasContextHandle = await page.evaluateHandle(canvas => canvas.getContext('2d'), canvasHandle);
```

Comment:
does this have issues when running headed vs headless with chrome (gpu)

Reply:
Possibly? The issue was that the element handles were being declared inside of `page.evaluate()` and these were sometimes being disposed too early, leading to failing tests. Declaring the handle outside of `page.evaluate()` ensures that they will be available for the entire lifetime of the test.